tower-http/src/services/fs/serve_dir.rs

```diff
@@ -192,6 +203,34 @@ impl<ReqBody> Service<Request<ReqBody>> for ServeDir {
     }
 }
 
+fn parse_range(headers: &HeaderMap) -> Option<ParsedRangeHeader> {
+    // ex:  `Range: bytes=0-1023`
+    let range_header = headers.get(header::RANGE)?;
+    if let Ok(as_str) = range_header.to_str() {
+        // produces ("bytes", "0-1023")
+        return if let Some((_, range)) = as_str.split_once("=") {
+            // produces ("0", "1023")
+            if let Some((start_incl, end_incl)) = range.split_once("-") {
+                if let Ok(start) = start_incl.parse() {
+                    if let Ok(end) = end_incl.parse() {
+                        if start < end {
+                            return Some(ParsedRangeHeader::Valid(ContentRange {
+                                start_inclusive: start,
+                                end_inclusive: end
+                            }));
+                        }
+                    }
+                }
+            }
+            // Range header at least on form bytes=<some_range> should produce a 416
+            Some(ParsedRangeHeader::Unsatisfiable)
+        } else {
+            Some(ParsedRangeHeader::Malformed("Range header value not in format `bytes=<range>`, `=` missing".to_owned()))
+        }
+    }
+    Some(ParsedRangeHeader::Malformed("Range not parseable as utf8".to_owned()))
+}
```

Comment:
Wondering if there is way to write this using the fact that `?` works on `Option` as well 🤔